DREAM/Optimization/tsgParticleSwarm.cpp

```diff
@@ -0,0 +1,168 @@
+                                 "number of dimensions (" + std::to_string(num_dimensions) + ")");
+    }
+    particle_positions = pp;
+}
+
+
+void ParticleSwarmState::setParticleVelocities(std::vector<double> &pv) {
+    if (pv.size() % num_dimensions != 0) {
+        throw std::runtime_error("Size of particle velocities (" + std::to_string(pv.size()) + ") is not a multiple of the " +
+                                 "number of dimensions (" + std::to_string(num_dimensions) + ")");
+    }
+    particle_velocities = pv;
+}
+
+
+void ParticleSwarmState::setBestParticlePositions(std::vector<double> &bpp) {
+    if (bpp.size() % num_dimensions != 0) {
+        throw std::runtime_error("Size of best particle positions (" + std::to_string(bpp.size()) + ") is not a multiple of the " +
+                                 "number of dimensions (" + std::to_string(num_dimensions) + ")");
+    }
+    best_particle_positions = bpp;
+}
+
+void ParticleSwarmState::setBestPosition(std::vector<double> &bp) {
+    if (bp.size() != (size_t) num_dimensions) {
+        throw std::runtime_error("Size of best position (" + std::to_string(bp.size()) + ") is not equal to the number of " +
+                                 "dimensions (" + std::to_string(num_dimensions) + ")");
+    }
+    best_position = bp;
+}
+
+void ParticleSwarmState::addParticlesInsideBox(int box_num_particles, std::vector<double> &box_lower, std::vector<double> &box_upper,
+                                               std::function<double(void)> get_random01) {
+    if (box_lower.size() != box_upper.size())
+        throw std::runtime_error("The size of box lower bounds and box upper bounds do not match");
+    if (num_dimensions == 0) num_dimensions = box_lower.size();
+    num_particles += box_num_particles;
+
+    std::vector<double> box_particle_positions(num_dimensions * box_num_particles);
+    std::generate(box_particle_positions.begin(), box_particle_positions.end(), get_random01);
+    for (int i=0; i<box_num_particles; i++) {
+        for (int j=0; j<num_dimensions; j++) {
+            box_particle_positions[i * num_dimensions + j] *= (box_upper[i] - box_lower[i]);
+            box_particle_positions[i * num_dimensions + j] -= box_lower[i];
+        }
+    }
+    std::copy(box_particle_positions.begin(), box_particle_positions.end(), std::back_inserter(particle_positions));
+
+    std::vector<double> box_particle_velocities(num_dimensions * box_num_particles);
+    std::generate(box_particle_velocities.begin(), box_particle_velocities.end(), get_random01);
+    std::copy(box_particle_velocities.begin(), box_particle_velocities.end(), std::back_inserter(particle_velocities));
+}
+
+void ParticleSwarm(BatchedObjectiveFunction f, int max_iterations, TasDREAM::DreamDomain inside, ParticleSwarmState &state,
+                   double inertia_weight, double cognitive_coeff, double social_coeff, std::function<double(void)> get_random01) {
```

Comment:
`get_random01` should always gave a default value, look at DREAM.

Reply:
Should we have to do this if default values are already given in the associated header file?

Reply:
The header is enough, in fact it won't compile if you set defaults twice (even if it is the same default).

I didn't see the header.